firmware/quadruna/VC/src/tasks.c

```diff
@@ -475,6 +515,44 @@ _Noreturn void tasks_runCanRx(void)
     }
 }
 
+_Noreturn void tasks_runLogging(void)
+{
+    static uint32_t count = 0;
+    for (;;)
+    {
+        if (!sd_functional())
+        {
+            osDelay(1000);
+            continue;
+        }
+
+        io_canLogging_recordMsgFromQueue();
+        count++;
+        write_count++;
+        if (count > 256)
+        {
+            io_canLogging_sync();
+            count = 0;
+        }
+    }
+}
+
+static void can_msg_received_callback(CanMsg *rx_msg)
+{
+    // TODO: check gpio present
+    io_can_msgReceivedCallback(rx_msg); // push to queue
+    if (sd_functional())
+    {
+        io_canLogging_msgReceivedCallback(rx_msg); // push to logging queue
+    }
+    read_count++;
+}
+
+static bool sd_functional(void)
+{
+    return !hw_gpio_readPin(&sd_present);
+}
```

Comment:
please just inline this function.

Reply:
or move it to io/app function. Keep tasks.c as minimal as possible.

Reply:
I don't want the io function coupled with this gpio pin. This is the reason I put it here.